pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/GroupedStages.java

```diff
@@ -0,0 +1,177 @@
+    String content = getGroups().stream()
+        .map(group ->
+            "[" + group.stream()
+                .map(stage -> Integer.toString(stage.getStageId()))
+                .collect(Collectors.joining(", ")) + "]"
+        )
+        .collect(Collectors.joining(", "));
+
+    return "[" + content + "]";
+  }
+
+  /**
+   * A mutable version of {@link GroupedStages}.
+   */
+  public static class Mutable extends GroupedStages {
+    /**
+     * All groups of stages.
+     *
+     * Although these groups are never empty, a group may contain only one stage if it is not grouped with any other
+     * stage.
+     */
+    private final SortedSet<SortedSet<MailboxSendNode>> _groups = new TreeSet<>(GROUP_COMPARATOR);
+
+    /**
+     * Map from stage to the group of stages it belongs to.
+     */
+    private final IdentityHashMap<MailboxSendNode, SortedSet<MailboxSendNode>> _stageToGroup = new IdentityHashMap<>();
+
+    /**
+     * Adds a new group of equivalent stages.
+     *
+     * @param node The stage that will be the only member of the group.
+     * @return this object
+     * @throws IllegalArgumentException if the stage was already added.
+     */
+    public Mutable addNewGroup(MailboxSendNode node) {
+      Preconditions.checkArgument(!containsStage(node), "Stage {} was already added", node.getStageId());
+      SortedSet<MailboxSendNode> group = new TreeSet<>(STAGE_COMPARATOR);
+      group.add(node);
+      _groups.add(group);
+      _stageToGroup.put(node, group);
+      return this;
+    }
+
+    /**
+     * Adds a stage to an existing group.
+     * @param original A stage that is already in the group.
+     * @param newNode The stage to be added to the group.
+     * @return this object
+     */
+    public Mutable addToGroup(MailboxSendNode original, MailboxSendNode newNode) {
+      Preconditions.checkArgument(!containsStage(newNode), "Stage {} was already added", newNode.getStageId());
+      SortedSet<MailboxSendNode> group = getGroup(original);
+      group.add(newNode);
+      _stageToGroup.put(newNode, group);
+      return this;
+    }
+
+    @Override
+    public SortedSet<MailboxSendNode> getLeaders() {
+      return _groups.stream()
+          .map(SortedSet::first)
+          .collect(Collectors.toCollection(() -> new TreeSet<>(STAGE_COMPARATOR)));
+    }
+
+    @Override
+    public SortedSet<SortedSet<MailboxSendNode>> getGroups() {
+      return _groups;
+    }
+
+    @Override
+    public boolean containsStage(MailboxSendNode stage) {
+      return _stageToGroup.containsKey(stage);
+    }
+
+    @Override
+    public SortedSet<MailboxSendNode> getGroup(MailboxSendNode stage)
+        throws NoSuchElementException {
+      SortedSet<MailboxSendNode> group = _stageToGroup.get(stage);
+      if (group == null) {
+        throw new NoSuchElementException("Stage " + stage.getStageId() + " is unknown by this class");
+      }
+      return group;
+    }
+
+    public Mutable removeStage(MailboxSendNode stage) {
```

Comment:
In the prune phase I was talking about before, although I'm not sure if that will be done in a copy fashion or actually modifying the object.

Basically in the next phase we are going to need to transforma a GroupedStages into a simplified expression tree where for each equivalent group `G` we create a new synthetic stage that send data to each stage `g` that belongs to `G`. But imagine a case like:

```mermaid
flowchart LR
    root --> S1
    S1 --> S2
    S1 --> S5
    S2 --> S3[S3 Read A]
    S2 --> S4[S4 Read B]
    S5 --> S6[S6 Read A]
    S5 --> S7[S7 Read B]
```

Where:
  - Stage 3 is equivalent to Stage 6
  - Stage 4 is equivalent to Stage 7
  - Stage 2 is equivalent to Stage 5

We wouldn't want to create synthetic classes for each of these equivalence groups because in the end all (S3, S6) and (S4, S7) could be pruned once (S2, S5) creates its own equivalence.

The final plan should be something like:
```mermaid
flowchart LR
    root --> S1
    S1 --> S2
    S1 --> S5
    S2 --> N1
    S5 --> N1

    N1 --> N2[N2 Read A]
    N1 --> N3[N2 Read B]
```

Where Ns are the new synthetic stages.

One way to implement this is to remove stages that will be pruned it its parent is equivalent. In our case S3, S4, S6 and S7 would be pruned.

I'm not 100% sure if this method will be called in the final, but at least in my first draft it is useful

Reply:
Good question. We need to keep them, but we also modify them. It is explained in the design document, but basically there are cases where we need to keep them. For example imagine that the sender node in S2 (the one that sends to S1) uses hash distribution while the sender of S5 sends using random. We could consider them not equivalent but it is better to just keep both stages, schedule them in a way they are running in the same servers than N1 and then apply each distribution.

Reply:
Wouldn't they still be replaced by new stages? The new ones will retain the sender nodes from the original but the rest of the stage will be replaced by a single receive node which gets data from the synthetic stage that contains all the other nodes (apart from the sender) from the equivalent stages right?

Reply:
Yes, S2 and S5 will change internally so we could consider them new as well. It helps me to think that they are not new but they have been modified, but this is just a concept level, not programatically (they will be different objects)